torch/csrc/autograd/variable.h

```diff
@@ -1,150 +1,255 @@
+
+  const char* toString() const override;
+  at::IntList sizes() const override;
+  at::IntList strides() const override;
+  int64_t dim() const override;
+  at::Scalar localScalar() override;
+  void* unsafeGetTH(bool retain) override;
+  std::unique_ptr<at::Storage> storage() override;
+  static const char* typeString();
+
   std::shared_ptr<Function> get_grad_accumulator();
-  virtual std::shared_ptr<Function>& get_grad_fn() { return _grad_fn; }
+  virtual std::shared_ptr<Function>& get_grad_fn() {
+    return grad_fn;
+  }
 
+  std::string name;
   at::Tensor data;
+
   Variable grad;
-  std::shared_ptr<Function> _grad_fn;
+  std::shared_ptr<Function> grad_fn;
+  std::weak_ptr<Function> grad_accumulator;
+
   VariableVersion version_counter;
   std::vector<std::shared_ptr<FunctionPreHook>> hooks;
-  std::weak_ptr<Function> grad_accumulator;
-  // Mutex to ensure that concurrent read operations that modify internal state
-  // are still thread-safe. Used by get_grad_fn and get_grad_accumulator.
-  std::mutex mutex;
-  bool _requires_grad;  // only meaningful on leaf variables (must be false otherwise)
-  bool is_view;
+
+  bool requires_grad; // only meaningful on leaf variables (must be false
+                      // otherwise)
+  bool is_view = false;
   // The "output number" of this variable; e.g., if this variable
   // was the second output of a function, then output_nr == 1.
   // We use this to make sure we can setup the backwards trace
   // correctly when this variable is passed to another function.
   int output_nr;
-  PyObject *pyobj;  // weak reference
+  PyObject* pyobj = nullptr; // weak reference
 
-  std::string name;
+  // Mutex to ensure that concurrent read operations that modify internal
+  // state are still thread-safe. Used by get_grad_fn and
+  // get_grad_accumulator.
+  std::mutex mutex;
 
   // For use in torch::jit::tracer
   auto_unique_ptr<jit::tracer::ValueTracingState> tracing_state;
-  friend struct VariableType;
 };
 
+//===----------------------------------------------------------------------===//
+//                          Variable::ViewImpl
+//===----------------------------------------------------------------------===//
+
 // A Variable that is a view on another Variable. The base and view share the
-// same version_counter. The _grad_fn field of the Variable may become stale
-// due to in-place modifications of the shared data. Accesses should go through
-// get_grad_fn(). All other fields are always valid.
-struct VariableViewImpl : public VariableImpl {
-  VariableViewImpl(Variable base, at::Tensor data, int output_nr, std::shared_ptr<Function> grad_fn);
+// same version_counter. The grad_fn field of the Variable may become stale
+// due to in-place modifications of the shared data. Accesses should go
+// through get_grad_fn(). All other fields are always valid.
+struct Variable::ViewImpl : public Variable::Impl {
+  ViewImpl(Variable base_, at::Tensor data_, Edge gradient_edge);
```

Comment:
Can we not call arguments with an underscore at the end? We use this convention for naming struct members in certain places, so let's avoid this

Reply:
But then in the constructor it's ambiguous if you have `: base(base)`?

Reply:
idk, is it? It's quite a common idiom

Reply:
Alternatively how about adding them at the beginning to avoid confusion?

Reply:
I renamed them to `base` (without underscores).

Reply:
This was a bad idea. The variables shadow each other in nasty ways. I'll put the underscores in the definition and not the declaration, if that helps.